src/java.base/share/classes/java/lang/reflect/code/type/ClassType.java

```diff
@@ -35,30 +40,47 @@
  */
 public final class ClassType implements TypeVarRef.Owner, JavaType {
     // Fully qualified name
-    private final String type;
+    private final ClassDesc type;
 
     private final List<JavaType> typeArguments;
 
-    ClassType(String type) {
+    ClassType(ClassDesc type) {
         this(type, List.of());
     }
 
-    ClassType(String type, List<JavaType> typeArguments) {
-        switch (type) {
-            case "boolean", "char", "byte", "short", "int", "long",
-                    "float", "double", "void" -> throw new IllegalArgumentException();
+    ClassType(ClassDesc type, List<JavaType> typeArguments) {
+        if (!type.isClassOrInterface()) {
+            throw new IllegalArgumentException("Invalid base type: " + type);
         }
         this.type = type;
         this.typeArguments = List.copyOf(typeArguments);
     }
 
+    @Override
+    public Type resolve(Lookup lookup) throws ReflectiveOperationException {
+        Class<?> baseType = type.resolveConstantDesc(lookup);
+        List<Type> resolvedTypeArgs = new ArrayList<>();
+        for (JavaType typearg : typeArguments) {
+            resolvedTypeArgs.add(typearg.resolve(lookup));
+        }
+        return resolvedTypeArgs.isEmpty() ?
+                baseType :
+                makeReflectiveParameterizedType(baseType,
+                        resolvedTypeArgs.toArray(new Type[0]), baseType.getDeclaringClass()); // @@@: generic owner is erased here
```

Comment:
Note: a `ParameterizedType` has the ability to model a "generic owner type". E.g. in the case of:
```
Foo<String>.Bar<Integer>
```
We currently don't support enclosing types/instances, so it's ok to just use the declaring class of the current type (which will be likely `null`).